spring-integration-core/src/main/java/org/springframework/integration/scattergather/ScatterGatherHandler.java

```diff
@@ -182,15 +178,27 @@ protected Object handleRequestMessage(Message<?> requestMessage) {
 				getMessageBuilderFactory()
 						.fromMessage(requestMessage)
 						.setHeader(GATHER_RESULT_CHANNEL, gatherResultChannel)
-						.setHeader(ORIGINAL_REPLY_CHANNEL, requestMessageHeaders.getReplyChannel())
-						.setHeader(ORIGINAL_ERROR_CHANNEL, requestMessageHeaders.getErrorChannel())
 						.setReplyChannel(this.gatherChannel)
 						.setErrorChannelName(this.errorChannelName)
 						.build();
 
 		this.messagingTemplate.send(this.scatterChannel, scatterMessage);
 
-		return gatherResultChannel.receive(this.gatherTimeout);
+		Message<?> gatherResult = gatherResultChannel.receive(this.gatherTimeout);
+
+		if (gatherResult != null) {
+			if (gatherResult instanceof ErrorMessage) {
+				throw (RuntimeException) gatherResult.getPayload();
```

Comment:
I don't think this is a framework responsibility to make such an unconditional decision.
In many other places we really just produce a returned message as is without any assumption about its content or type.
It would be much better to let end-user to make a logic from such a returned message.

Reply:
HI @artembilan ,

Good Morning!!

I understand it is not framework responsibility to handle functional scenario's. But here it is behaviour of the ScatterGather that needs to be consistent. when outputProcessor throws an exception, ScatterGather needs to propagate it to proper Error Channel.

When output processor throws exception, the exception will be wrapped in Error message, it will be propagated to ScatterGatherResultChannel. This message will be propagated to next channel in the FlowDefinition.

Please have a look at the flow in the testcase propagateErrorFromGatherer.

`@Bean
		public IntegrationFlow propagateErrorFromGatherer(TaskExecutor taskExecutor) {
			return IntegrationFlows.from(Function.class)
					.scatterGather(s -> s
									.applySequence(true)
									.recipientFlow(subFlow -> subFlow
											.channel(c -> c.executor(taskExecutor))
											.transform(p -> "foo")),
							g -> g
									.outputProcessor(group -> {
										throw new RuntimeException("intentional");
									}),
							sg -> sg.gatherTimeout(100))
					.transform(m -> "This should not be executed, results must have been propagated to Error Channel")
					.get();
		}`

When `.transform(m -> "This should not be executed, results must have been propagated to Error Channel")` was not added, test case passed with success as it was able to get the Expected error response from the Error Channel.

But same test case failed when the .transform is added. This behavior needs to be corrected by handling the Exception. Similar checks are performed in the class MessagingGatewaySupport.java. when reply message is received, it is checked for instance of ErrorMessage or Throwable class

`	@Nullable // NOSONAR
	private Object doSendAndReceive(Object object, boolean shouldConvert) {
		initializeIfNecessary();
		Assert.notNull(object, "request must not be null");
		MessageChannel channel = getRequestChannel();
		if (channel == null) {
			throw new MessagingException("No request channel available. Cannot send request message.");
		}

		registerReplyMessageCorrelatorIfNecessary();

		Object reply;
		Message<?> requestMessage = null;
		try {
			if (this.countsEnabled) {
				this.messageCount.incrementAndGet();
			}
			if (shouldConvert) {
				reply = this.messagingTemplate.convertSendAndReceive(channel, object, Object.class,
						this.historyWritingPostProcessor);
			}
			else {
				requestMessage = (object instanceof Message<?>)
						? (Message<?>) object : this.requestMapper.toMessage(object);
				Assert.state(requestMessage != null, () -> "request mapper resulted in no message for " + object);
				requestMessage = this.historyWritingPostProcessor.postProcessMessage(requestMessage);
				reply = this.messagingTemplate.sendAndReceive(channel, requestMessage);
			}

			if (reply == null && this.errorOnTimeout) {
				throwMessageTimeoutException(object, "No reply received within timeout");
			}
		}
		catch (Exception ex) {
			if (logger.isDebugEnabled()) {
				logger.debug("failure occurred in gateway sendAndReceive: " + ex.getMessage());
			}
			reply = ex;
		}

		if (reply instanceof Throwable || reply instanceof ErrorMessage) {
			Throwable error =
					reply instanceof ErrorMessage
							? ((ErrorMessage) reply).getPayload()
							: (Throwable) reply;
			return handleSendAndReceiveError(object, requestMessage, error, shouldConvert);
		}
		return reply;
	}
`

Reply:
Well, this is definitely not a `gateway` here. We check there for `ErrorMessage` because a high-level contract is a plain old Java method invocation, so we need to honor its signature.

Here we talk about internals of Spring Integration which deal with messages. I fully not agree to re-throw that exception. If you return an `ErrorMessage` from the gatherer, it is already up to your downstream flow to process it this or other way. The logic has to be a simple as possible. We should not mix concerns in one component. We have hidden already the maximum, but that's not our responsibility to make a decision about returned message. We only need to correct headers nothing more. Even if it is not an `ErrorMessage`, you still can return an `Exception` in the `payload`. The flow you are mentioned really has to be mentioned with a `handle()` or `filter()` to decide what to do with that exception. That's not what we want to do in the `Scatter-Gather` hard logic.

Reply:
Hi @artembilan 

If you see the implemetation of the AbstractCorrelatingMessageHandler (gather handler), when the outprocessor throws exception this will be wrapped in the error message and sent to Error Channel.

Now if we don't throw error in the handleRequestMessage, message will flow to the next channel in the flow instead of being raised as exception.